src/custom/hooks/useTokenLazy.ts

```diff
@@ -0,0 +1,148 @@
+import { useCallback } from 'react'
+import { Web3Provider } from '@ethersproject/providers'
+
+import { Token } from '@uniswap/sdk-core'
+import { Contract } from '@ethersproject/contracts'
+
+import { useActiveWeb3React } from 'hooks/web3'
+import { getBytes32TokenContract, getTokenContract } from 'hooks/useContract'
+import { parseStringOrBytes32 } from 'hooks/Tokens'
+import { useAddUserToken } from 'state/user/hooks'
+import { Erc20 } from 'abis/types'
+import { retry } from 'utils/retry'
+
+const contractsCache: Record<string, Erc20> = {}
+const bytes32ContractsCache: Record<string, Contract> = {}
+
+const RETRY_OPTIONS = { n: 3, minWait: 100, maxWait: 3_000 }
+
+export interface TokenInfo {
+  decimals: number
+  symbol?: string
+  name?: string
+}
+
+export interface GetTokenInfoParams {
+  chainId: number
+  address: string
+  account: string
+
+  library: Web3Provider
+}
+
+async function _getTokenContract(params: GetTokenInfoParams): Promise<Erc20> {
+  const { address, account, chainId, library } = params
+
+  return contractsCache[address] || getTokenContract(address, undefined, library, account, chainId)
+}
+
+async function _getTokenByte32Contract(params: GetTokenInfoParams): Promise<Contract> {
+  const { address, account, chainId, library } = params
+
+  return bytes32ContractsCache[address] || getBytes32TokenContract(address, undefined, library, account, chainId)
+}
+
+async function _getTokenInfo(params: GetTokenInfoParams): Promise<TokenInfo | null> {
```

Comment:
this one i feel is too big, but i think we can leave it like this for now...
I also want this PR to be merged :)